flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSerializer.java

```diff
@@ -170,7 +170,11 @@ public int getLength() {
 
 	@Override
 	public T createInstance() {
-		return InstantiationUtil.instantiate(type);
+		if (!isGenericRecord(type)) {
```

Comment:
Absolutely! This also remove any use of InstantiationUtil that does not seem to have any support for any AVRO idiosyncrasy

Reply:
Sorry, but this solution causes many tests to fail:
```
[ERROR] Failures:
[ERROR]   AvroGenericTypeComparatorTest>AbstractGenericTypeComparatorTest.testBeanStyleObjects:78->AbstractGenericTypeComparatorTest.runTests:101 Exception in test: null
[ERROR]   AvroGenericTypeComparatorTest>AbstractGenericTypeComparatorTest.testCompositeObject:65->AbstractGenericTypeComparatorTest.runTests:101 Exception in test: null
[ERROR]   AvroGenericTypeComparatorTest>AbstractGenericTypeComparatorTest.testSimpleTypesObjects:45->AbstractGenericTypeComparatorTest.runTests:101 Exception in test: null
[ERROR]   AvroGenericTypeComparatorTest>AbstractGenericTypeComparatorTest.testString:33->AbstractGenericTypeComparatorTest.runTests:101 Exception in test: null
[ERROR]   AvroGenericTypeSerializerTest>AbstractGenericTypeSerializerTest.testBeanStyleObjects:95->AbstractGenericTypeSerializerTest.runTests:155 Exception in test: null
[ERROR]   AvroGenericTypeSerializerTest>AbstractGenericTypeSerializerTest.testCompositeObject:75->AbstractGenericTypeSerializerTest.runTests:155 Exception in test: null
[ERROR]   AvroGenericTypeSerializerTest>AbstractGenericTypeSerializerTest.testNestedInterfaces:124->AbstractGenericTypeSerializerTest.runTests:155 Exception in test: null
[ERROR]   AvroGenericTypeSerializerTest>AbstractGenericTypeSerializerTest.testNestedObjects:85->AbstractGenericTypeSerializerTest.runTests:155 Exception in test: null
[ERROR]   AvroGenericTypeSerializerTest>AbstractGenericTypeSerializerTest.testSimpleTypesObjects:64->AbstractGenericTypeSerializerTest.runTests:155 Exception in test: null
[ERROR]   AvroGenericTypeSerializerTest>AbstractGenericTypeSerializerTest.testString:41->AbstractGenericTypeSerializerTest.runTests:155 Exception in test: null
[ERROR]   AvroSerializerEmptyArrayTest.testBookSerialization:42 Exception in test: null
[ERROR]   AvroSerializerEmptyArrayTest.testSerialization:67 Exception in test: null
[ERROR]   AvroSerializerTest>SerializerTestBase.testCopyIntoNewElements:207 Exception in test: null
[ERROR]   AvroSerializerTest>SerializerTestBase.testCopyIntoReusedElements:229 Exception in test: null
[ERROR]   AvroSerializerTest>SerializerTestBase.testInstantiate:111 Exception in test: null
[ERROR]   AvroSerializerTest>SerializerTestBase.testSerializeAsSequenceReusingValues:354 Exception in test: null
[ERROR]   AvroSerializerTest>SerializerTestBase.testSerializeIndividually:257 Exception in test: null
[ERROR]   AvroSerializerTest>SerializerTestBase.testSerializeIndividuallyReusingValues:290 Exception in test: null
[ERROR]   AvroSerializerTest>SerializerTestBase.testSerializedCopyAsSequence:425 Exception in test: null
[ERROR]   AvroSerializerTest>SerializerTestBase.testSerializedCopyIndividually:387 Exception in test: null
```

Reply:
Not it's more consistent with other methods but still using `InstantiationUtil.instantiate()` for specific records. They are concrete and have default constructor, so it should be ok

Reply:
Sorry, I have not run the code I posted before and I used the wrong field for schema. I think we can use the approach from the beginning of the discussion if we use the `runtimeSchema`:

```
	@Override
	@SuppressWarnings("unchecked")
	public T createInstance() {
		checkAvroInitialized();
		return (T) avroData.newRecord(null, runtimeSchema);
	}
```

The `schema` field may not always be set as it is used for checkpointing and for `Specific/ReflectData` it will not be set from the beginning. The `runtimeSchema` is always set though.